src/solvers/dg/2d/amr.jl

```diff
@@ -11,21 +11,46 @@ function refine!(dg::Dg2D{Eqn, MeshType, NVARS, POLYDEG}, mesh::TreeMesh,
   # Determine for each existing element whether it needs to be refined
   needs_refinement = falses(nelements(dg.elements))
   tree = mesh.tree
-  # The "Ref(...)" is such that we can vectorize the search but not the array that is searched
-  elements_to_refine = searchsortedfirst.(Ref(dg.elements.cell_ids[1:nelements(dg.elements)]),
-                                          cells_to_refine)
+
+  # Find all indices of elements whose cell ids are in cells_to_refine
+  elements_to_refine = findall(cell_id -> cell_id in cells_to_refine, dg.elements.cell_ids)
+
   needs_refinement[elements_to_refine] .= true
 
   # Retain current solution data
   old_n_elements = nelements(dg.elements)
   old_u = dg.elements.u
 
-  # Get new list of leaf cells
-  leaf_cell_ids = leaf_cells(tree)
-
-  # Initialize new elements container
-  elements = init_elements(leaf_cell_ids, mesh, Val(NVARS), Val(POLYDEG))
-  n_elements = nelements(elements)
+  (elements, n_elements, interfaces, n_interfaces, mpi_interfaces, n_mpi_interfaces,
+      boundaries, n_boundaries, n_boundaries_per_direction, mortar_type, l2mortars, 
+      ecmortars, n_l2mortars, n_ecmortars, mpi_neighbor_ranks, 
+      mpi_neighbor_interfaces, mpi_send_buffers, mpi_recv_buffers, 
+      mpi_send_requests, mpi_recv_requests, n_elements_by_rank, n_elements_global, 
+      first_element_global_id) = initialize_containers(mesh, ndims(dg), NVARS, POLYDEG)
+  
+  # Update DG instance with new data
+  dg.elements = elements
+  dg.n_elements = n_elements
+  dg.interfaces = interfaces
+  dg.n_interfaces = n_interfaces
+  dg.mpi_interfaces = mpi_interfaces
+  dg.n_mpi_interfaces = n_mpi_interfaces
+  dg.boundaries = boundaries
+  dg.n_boundaries = n_boundaries
+  dg.n_boundaries_per_direction = n_boundaries_per_direction
+  dg.l2mortars = l2mortars
+  dg.n_l2mortars = n_l2mortars
+  dg.ecmortars = ecmortars
+  dg.n_ecmortars = n_ecmortars
+  dg.mpi_neighbor_ranks = mpi_neighbor_ranks
+  dg.mpi_neighbor_interfaces = mpi_neighbor_interfaces
+  dg.mpi_send_buffers = mpi_send_buffers
+  dg.mpi_recv_buffers = mpi_recv_buffers
+  dg.mpi_send_requests = mpi_send_requests
+  dg.mpi_recv_requests = mpi_recv_requests
+  dg.n_elements_by_rank = n_elements_by_rank
+  dg.n_elements_global = n_elements_global
+  dg.first_element_global_id = first_element_global_id
 
   # Loop over all elements in old container and either copy them or refine them
   element_id = 1
```

Comment:
It sounds wrong to me to first initialize all containers (interfaces, boundaries, ...) and then copy/refine the elements, but it works like this. Is this okay, @sloede?

Reply:
I don't think we'll be able to use such a structure when moving to Taal (see #200), so l'd say leave it for now as it is, and revisit this later when porting your code to Taal.